src/behaviours/use-drag.ts

```diff
@@ -0,0 +1,584 @@
+}
+
+const getConstraints = (
+    axis: "x" | "y",
+    { top, right, bottom, left }: Constraints
+) => {
+    if (axis === "x") {
+        return { min: left, max: right }
+    } else {
+        return { min: top, max: bottom }
+    }
+}
+
+function applyConstraints(
+    axis: "x" | "y",
+    value: number | MotionValue<number>,
+    constraints: Constraints | false,
+    dragElastic: boolean | number
+): number {
+    let constrainedValue = value instanceof MotionValue ? value.get() : value
+    if (!constraints) {
+        return constrainedValue
+    }
+    const { min, max } = getConstraints(axis, constraints)
+
+    if (min !== undefined && constrainedValue < min) {
+        constrainedValue = dragElastic
+            ? applyOverdrag(min, constrainedValue, dragElastic)
+            : Math.max(min, constrainedValue)
+    } else if (max !== undefined && constrainedValue > max) {
+        constrainedValue = dragElastic
+            ? applyOverdrag(max, constrainedValue, dragElastic)
+            : Math.min(max, constrainedValue)
+    }
+    if (value instanceof MotionValue) {
+        value.set(constrainedValue)
+    }
+    return constrainedValue
+}
+
+function applyOverdrag(
+    origin: number,
+    current: number,
+    dragElastic: boolean | number
+) {
+    const dragFactor = typeof dragElastic === "number" ? dragElastic : 0.35
+    return mix(origin, current, dragFactor)
+}
+
+function bothAxis<T>(handler: (axis: "x" | "y") => T): T[] {
+    return [handler("x"), handler("y")]
+}
+
+type MotionPoint = Partial<{
+    x: MotionValue<number>
+    y: MotionValue<number>
+}>
+
+interface DragState {
+    isDragging: boolean
+    hasDragged: boolean
+    currentDirection: DragDirection | null
+    constraints: Constraints | false
+    handlers: DragHandlers
+}
+
+/**
+ * A hook that allows an element to be dragged.
+ *
+ * @param param
+ * @param ref
+ * @param values
+ * @param controls
+ *
+ * @internal
+ */
+export function useDrag(
+    {
+        drag = false,
+        dragDirectionLock = false,
+        dragPropagation = false,
+        dragConstraints = false,
+        dragElastic = true,
+        dragMomentum = true,
+        _dragValueX,
+        _dragValueY,
+        _dragTransitionControls,
+        dragOriginX,
+        dragOriginY,
+        dragTransition,
+        onDirectionLock,
+        onDragStart,
+        onDrag,
+        onDragEnd,
+        onDragTransitionEnd,
+    }: DraggableProps,
+    ref: RefObject<Element>,
+    values: MotionValuesMap,
+    controls: ValueAnimationControls
+) {
+    // If `dragConstraints` is a React `ref`, we should resolve the constraints once the
+    // component has rendered.
+    const constraintsNeedResolution = isRefObject(dragConstraints)
+    const state = useRef<DragState>({
+        isDragging: false,
+        hasDragged: false,
+        currentDirection: null,
+        constraints: false,
+        handlers: {},
+    }).current
+
+    // TODO: Explain why we keep these in a ref
```

Comment:
Let's turn this TODO into a comment explaining exactly that 😄

Reply:
It seems like if this is not react state created by React.useState, it should not be called state